ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+		for observer := range c.observers {
+			observer.OnComplete()
+		}
+
+		// Signal Compactor goroutine to stop
+		doneCh := make(chan struct{})
+		c.stopCh <- doneCh
+
+		// Wait for Compactor goroutine to stop
+		<-doneCh
+
+		// Close trieUpdateDoneCh to signal trie updates are finished
+		close(c.trieUpdateDoneCh)
+	})
+	return c.lm.Stopped()
+}
+
+func (c *Compactor) run() {
+
+	// checkpointSem is used to limit checkpointing to one.
+	// If previous checkpointing isn't finished when enough segments
+	// are finalized for next checkpointing, retry checkpointing
+	// again when next segment is finalized.
+	// This avoids having more tries in memory than needed.
+	checkpointSem := semaphore.NewWeighted(1)
+
+	checkpointResultCh := make(chan checkpointResult, 1)
+
+	// Get active segment number.
+	// activeSegmentNum is updated when record is written to a new segment.
+	_, activeSegmentNum, err := c.wal.Segments()
+	if err != nil {
+		c.logger.Error().Err(err).Msg("compactor failed to get active segment number")
+		activeSegmentNum = -1
+	}
+
+	lastCheckpointNum, err := c.checkpointer.LatestCheckpoint()
+	if err != nil {
+		c.logger.Error().Err(err).Msg("compactor failed to get last checkpoint number")
+		lastCheckpointNum = -1
+	}
+
+	// Compute next checkpoint number.
+	// nextCheckpointNum is updated when
+	// - checkpointing starts, fails to start, or fails.
+	// - tries snapshot fails.
+	// NOTE: next checkpoint number must >= active segment num.
+	// We can't reuse mtrie state to checkpoint tries in older segments.
+	nextCheckpointNum := lastCheckpointNum + int(c.checkpointDistance)
+	if activeSegmentNum > nextCheckpointNum {
+		nextCheckpointNum = activeSegmentNum
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+Loop:
+	for {
+		select {
+
+		case doneCh := <-c.stopCh:
+			defer close(doneCh)
+			cancel()
+			break Loop
+
+		case checkpointResult := <-checkpointResultCh:
+			if checkpointResult.err != nil {
+				c.logger.Error().Err(checkpointResult.err).Msgf(
+					"compactor failed to checkpoint %d", checkpointResult.num,
+				)
+
+				// Retry checkpointing after active segment is finalized.
+				nextCheckpointNum = activeSegmentNum
+			}
+
+		case update, ok := <-c.trieUpdateCh:
+			if !ok {
+				// trieUpdateCh channel is closed.
+				// Wait for stop signal from c.stopCh
+				continue
+			}
+
+			var checkpointNum int
+			var checkpointTries []*trie.MTrie
+			activeSegmentNum, checkpointNum, checkpointTries =
+				c.processTrieUpdate(update, activeSegmentNum, nextCheckpointNum)
+
+			if checkpointTries == nil {
+				// Don't checkpoint yet because
+				// - not enough segments for checkpointing (nextCheckpointNum >= activeSegmentNum), or
+				// - failed to get ledger state snapshop (nextCheckpointNum < activeSegmentNum)
+				if nextCheckpointNum < activeSegmentNum {
+					nextCheckpointNum = activeSegmentNum
+				}
+				continue
+			}
+
+			// Try to checkpoint
+			if checkpointSem.TryAcquire(1) {
```

Comment:
If a second call is trying to acquire before the first call release it, would the second call get blocked and get resumed after the first call is done?

Reply:
No, the second call would return without blocking.  If it fails to acquire semaphore, it retries checkpointing when active segment is finalized.

Reply:
I see. The lock is created outside of the for-loop. 

If the first call finishes before the second call, then the second call will trigger a new checkpoint.
If the first call hasn't finished before the second call, then the second call will not trigger a new checkpoint. After the first call finishes, it will also not trigger checkpointing. If the first call finishes before the third call, then the third call will trigger the second checkpoint. 

Right?